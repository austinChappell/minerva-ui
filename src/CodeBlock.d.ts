/// <reference types="react" />
export declare const liveEditorStyle: {
  fontSize: number;
  marginBottom: number;
  marginTop: number;
  fontFamily: string;
  borderRadius: number;
  padding: number;
};
export declare const liveErrorStyle: {
  fontFamily: string;
  fontSize: number;
  padding: string;
  color: string;
  backgroundColor: string;
};
declare const CodeBlock: {
  ({ className, live, isManual, render, children, ...props }: any): JSX.Element;
  defaultProps: {
    mountStylesheet: boolean;
  };
  displayName: string;
  __docgenInfo: {
    description: string;
    displayName: string;
    props: {};
  };
};

declare module '@mdx-js/react';

export default CodeBlock;
