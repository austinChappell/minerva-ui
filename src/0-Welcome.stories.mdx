<Meta title="About|Introduction" />

# Get started

Minerva UI is a reusable component library to help build UIs faster. This library aims to be highly composable, declarative and accessible.

# Install
`npm install --save minerva-ui`
or
`yarn add minerva-ui`

# Usage

Import components you want into your UI:

```js
import { Checkbox } from 'minerva-ui';
```

And use them:

```jsx
() => {
  const [checked, setChecked] = React.useState(false);
  return (
    <Checkbox
      checked={checked}
      onChange={() => setChecked(!checked)}
    >
      Stay Logged In
    </Checkbox>
  )
}
```

# Utility Props

Utility props are provided as aliases for most components. The style is heavily influenced by [Tailwind CSS](https://tailwindcss.com/docs/font-size).

For example:
```jsx
() => (
  <>
    {/* enter a custom pixel value */}
    <Button fontSize="12px">Save</Button>
    {/* or enter the value named "sm" in the theme, which is .875rem by default */}
    <Button fontSize="lg">Save</Button>
  </>
)
```

# Styling Props Example

Here's the ["Card" example](https://tailwindcss.com/components/cards) by Tailwind recreated using this API:

```jsx
() => (
  <Flex
    flexDirection="column"
    boxShadow="lg"
    borderRadius="md"
    maxWidth="24rem"
    overflow="hidden"
  >
    <img
      src="https://tailwindcss.com/img/card-top.jpg"
      alt="Sunset in the mountains"
    />
    <Flex flexDirection="column" px={6} py={4}>
      <Text fontWeight="bold" fontSize="xl" mb={2} color="gray.700">The Coldest Sunset</Text>
      <Text color="gray.700" lineHeight="normal">Lorem ipsum dolor sit amet, consectetur adipisicing elit. Voluptatibus quia, nulla! Maiores et perferendis eaque, exercitationem praesentium nihil.</Text>
    </Flex>
    <Flex px={6} py={4}>
      <Block color="gray.700" bg="gray.200" borderRadius="full" px={3} py={1} mr={2}>
        #photography
      </Block>
      <Block color="gray.700" bg="gray.200" borderRadius="full" px={3} py={1} mr={2}>
        #travel
      </Block>
      <Block color="gray.700" bg="gray.200" borderRadius="full" px={3} py={1}>
        #winter
      </Block>
    </Flex>
  </Flex>
)
```
<!-- ### Styling
To include base styling, you can import the styles for the corresponding component:
```
import 'minerva-ui/checkbox.css'
``` -->