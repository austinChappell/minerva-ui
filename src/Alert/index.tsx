import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { MinervaProps, systemProps, Flex } from '../layout';
import Icon from '../Icon';

const StyledAlert = styled(Flex)(
  props => ({
    padding: '16px',
    borderRadius: '5px',
  }),
  systemProps
);

const StyledAlertTitle = styled('p')(
  {
    fontWeight: 'bold',
    margin: '0 5px 0 8px',
  },
  systemProps
);

const StyledAlertDescription = styled.p``;

export interface AlertProps extends MinervaProps {
  children?: React.ReactNode;
  status?: 'error' | 'success' | 'warning' | 'info';
  title?: string;
  description?: string;
  color?: string;
  icon?: string;
  props?: any;
}

// @TODO: Move these to the theme so they can be customized
const alertTypes = {
  error: {
    statusColor: '#f8b4b4',
    statusIcon: 'alert-circle',
  },
  success: {
    statusColor: '#bcf0da',
    statusIcon: 'check-circle',
  },
  warning: {
    statusColor: '#fdf6b2',
    statusIcon: 'alert-triangle',
  },
  info: {
    statusColor: '#c3ddfd',
    statusIcon: 'info',
  },
};

export const Alert = forwardRef(function Alert(
  { title, description, status, color, icon, ...props }: AlertProps,
  ref
) {
  const { statusColor, statusIcon } =
    // if status is provided and is a provided alertType, return it
    status && Object.keys(alertTypes).includes(status)
      ? alertTypes[status]
      : {
          statusColor: '#e5e7eb',
          statusIcon: null,
        };

  const alertIcon = icon || statusIcon;

  return (
    <StyledAlert
      ref={ref}
      backgroundColor={color ? color : statusColor}
      {...props}
    >
      {alertIcon && <Icon name={alertIcon} size="20px" />}
      <StyledAlertTitle>{title}</StyledAlertTitle>
      <StyledAlertDescription>{description}</StyledAlertDescription>
    </StyledAlert>
  );
});

export default Alert;
