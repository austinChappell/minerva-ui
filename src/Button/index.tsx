import React, { forwardRef } from 'react';
import Spinner from '../Spinner';
import PseudoBox, { PseudoBoxProps } from '../PseudoBox';
import { MinervaProps, systemProps } from '../layout';
// import { useTheme } from '../theme';
import styled from 'styled-components';

const StyledButton = styled(PseudoBox)(
  props => ({
    ...props.theme.Button,
  }),
  systemProps
);

export interface ButtonProps extends MinervaProps, PseudoBoxProps {
  children?: React.ReactNode;
  /** Toggles disabled pseudo class */
  disabled?: boolean;
  /** If `true`, button will show a spinner. */
  isLoading?: boolean;
}

export const Button = forwardRef(function Button(
  {
    children,
    disabled = false,
    as: Comp = 'button',
    isLoading = false,
    ...props
  }: ButtonProps,
  ref
) {
  // const theme = useTheme();

  return (
    <StyledButton
      ref={ref}
      as={Comp}
      disabled={disabled || isLoading}
      role="button"
      transition="all 150ms ease 0s"
      outline="none"
      _hover={{
        backgroundColor: '#f9fafb',
      }}
      _focus={{
        borderColor: '#a4cafe',
        boxShadow: '0 0 0 3px rgba(118,169,250,.45)',
        outline: 0,
      }}
      _disabled={{
        opacity: 0.4,
        cursor: 'not-allowed',
      }}
      // {...theme.Button}
      {...props}
    >
      {isLoading ? <Spinner /> : children}
    </StyledButton>
  );
});

export default Button;
