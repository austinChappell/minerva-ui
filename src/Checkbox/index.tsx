import React from 'react';
import styled from 'styled-components';
import {
  background,
  border,
  color,
  flexbox,
  grid,
  layout,
  position,
  shadow,
  space,
  typography,
} from 'styled-system';
import { CustomCheckboxContainer, CustomCheckboxInput } from '@reach/checkbox';
import { MinervaProps } from '../layout';

/**
 * TODO:
 * - Add sizes
 */

const CheckboxContainer = styled.label({
  display: 'flex',
  alignItems: 'center',
  cursor: 'pointer',
  fontSize: '14px',
});

export type ControlBoxProps = {
  checked?: boolean;
  theme?: any;
};

const ControlBox = styled('div')<ControlBoxProps>(
  (props: ControlBoxProps) => ({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    borderStyle: 'solid',
    borderWidth: '1px',
    borderRadius: '4px',
    width: '16px',
    height: '16px',
    bg: 'primary',
    marginRight: '8px',
    transition: 'all 200ms ease',
    borderColor: props.checked ? '#fff' : '#ecebed',
    backgroundImage: `url("data:image/svg+xml;charset=utf-8,%3Csvg viewBox='0 0 16 16' fill='%23fff' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M5.707 7.293a1 1 0 00-1.414 1.414l2 2a1 1 0 001.414 0l4-4a1 1 0 00-1.414-1.414L7 8.586 5.707 7.293z'/%3E%3C/svg%3E")`,
    // backgroundColor: props.checked
    //   ? props.theme.colors.logoColor
    //   : '0px 7px 20px rgba(0, 0, 0, 0.00)',
    backgroundColor: props.checked ? '#5850ec' : '#fff',
    // boxShadow: props.checked
    //   ? '0px 7px 20px rgba(0, 0, 0, 0.07)'
    //   : '0px 7px 20px rgba(0, 0, 0, 0.00)',
    ':focus': {
      borderColor: '#a4cafe',
      boxShadow: '0 0 0 3px rgba(118,169,250,.45)',
      outline: 0,
    },
  }),
  color,
  space,
  flexbox,
  grid,
  layout,
  position,
  shadow,
  background,
  border,
  typography
);

// const CheckboxIcon = ({ checked = false, fill = '#fff' }) => {
//   return (
//     <svg
//       width="12"
//       height="12"
//       viewBox="0 0 16 16"
//       fill="none"
//       xmlns="http://www.w3.org/2000/svg"
//     >
//       <path
//         d="M11.4937 4.75844C11.8383 4.41385 12.397 4.41385 12.7416 4.75844C13.0861 5.10302 13.0861 5.66169 12.7416 6.00627L7.44745 11.3004C7.10287 11.645 6.54419 11.645 6.19961 11.3004L3.25844 8.35921C2.91385 8.01463 2.91385 7.45596 3.25844 7.11138C3.60302 6.7668 4.16169 6.7668 4.50627 7.11138L6.82353 9.42863L11.4937 4.75844Z"
//         fill={checked ? fill : 'transparent'}
//       />
//     </svg>
//   );
// };

const VisuallyHidden = styled.div`
  border: 0px;
  clip: rect(0px, 0px, 0px, 0px);
  height: 1px;
  width: 1px;
  padding: 0px;
  overflow: hidden;
  white-space: nowrap;
  position: absolute;
`;

type BaseProps = MinervaProps & React.InputHTMLAttributes<HTMLInputElement>;

// extend the native HTML attributes for nicer autocompletion
export interface CheckboxProps extends BaseProps {
  // define the custom props we're going to be using
  children?: React.ReactNode;
  // add a question mark to make it an optional prop
  disabled?: boolean;
  checked?: boolean;
  onChange?: () => any;
  style?: any;
}

export default function Checkbox(props: CheckboxProps) {
  const { children, checked = false, onChange, ...rest } = props;

  return (
    <CheckboxContainer>
      <CustomCheckboxContainer checked={checked} onChange={onChange}>
        <VisuallyHidden>
          <CustomCheckboxInput {...rest} />
        </VisuallyHidden>
        <ControlBox tabIndex={0} checked={checked} />
        {/* <CheckboxIcon checked={checked} /> */}
        {/* </ControlBox> */}
      </CustomCheckboxContainer>
      {children}
    </CheckboxContainer>
  );
}
