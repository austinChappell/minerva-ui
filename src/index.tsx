/**
 * Utilities
 */

export { default as ThemeProvider } from './ThemeProvider';
export { default as GlobalStyles } from './GlobalStyles';
export { default as defaultTheme } from './theme';
export { default as styled } from 'styled-components';

/**
 * Display
 */
export { default as Text } from './Text';
export { default as Link } from './Link';
export { default as Heading } from './Heading';
export { default as Tag } from './Tag';
export { default as Alert } from './Alert';

/**
 * Layout
 */
export { Box, Block, Flex } from './layout';
export { default as PseudoBox } from './PseudoBox';
export { default as Stack } from './Stack';

/**
 * Inputs
 */
export { default as Button } from './Button';
export { default as Checkbox } from './Checkbox';
export { default as Input } from './Input';
export { RadioGroup, Radio } from './Radio';
export { default as InputField } from './InputField';
export { default as Select } from './Select';
export { default as TagsInput } from './TagsInput';

/**
 * Elements
 */
export { default as Image } from './Image';
export { default as Table } from './Table';
export { default as Icon } from './Icon';
export * from './Table';
export { default as Skeleton } from './Skeleton';

/**
 * Helpers
 */
export { default as Spinner } from './Spinner';
