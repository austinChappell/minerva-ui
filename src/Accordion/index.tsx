import React, { forwardRef } from 'react';
import styled from 'styled-components';
import {
  Accordion as ReachAccordion,
  AccordionItem as ReachAccordionItem,
  AccordionButton as ReachAccordionButton,
  AccordionPanel as ReachAccordionPanel,
  AccordionProps as ReachAccordionProps,
} from '@reach/accordion';
import { MinervaProps, Box, systemProps } from '../layout';
import Icon from '../Icon';
import Button from '../Button';

export interface AccordionProps extends MinervaProps {
  children?: React.ReactNode;
  props?: any;
}

export interface AccordionItemProps extends MinervaProps {
  children?: React.ReactNode;
  props?: any;
}

export interface AccordionButtonProps extends MinervaProps {
  children?: React.ReactNode;
  props?: any;
}

export interface AccordionPanelProps extends MinervaProps {
  children?: React.ReactNode;
  props?: any;
}

type AllAccordionProps = ReachAccordionProps & MinervaProps;
export const Accordion = styled(ReachAccordion)<AllAccordionProps>(
  {
    borderBottomColor: 'gray.200',
    borderBottomWidth: '1px',
  },
  systemProps
);

export const AccordionItem = forwardRef(function AccordionItem(
  { ...props }: AccordionItemProps,
  ref
) {
  return <Box as={ReachAccordionItem} ref={ref} {...props} />;
});

export const AccordionButton = forwardRef(function AccordionButton(
  { children, ...props }: AccordionButtonProps,
  ref
) {
  return (
    <Button
      as={ReachAccordionButton}
      borderColor="transparent"
      width="100%"
      flexDirection="row"
      justifyContent="space-between"
      alignItems="flex-start"
      bg="white"
      py="12px"
      fontSize="18px"
      fontWeight="400"
      borderTopColor="gray.200"
      borderTopWidth="1px"
      ref={ref}
      {...props}
    >
      {children}
      {/* TODO Need to figure out how to determine if Item is expanded and rotate icon */}
      {/* <Icon name="chevron-down" ml={2} size="14px" /> */}
    </Button>
  );
});

export const AccordionPanel = forwardRef(function AccordionPanel(
  { ...props }: AccordionPanelProps,
  ref
) {
  return (
    <Box
      as={ReachAccordionPanel}
      px="14px"
      py="14px"
      width="95%"
      ref={ref}
      {...props}
    />
  );
});

export const MainAccordion = forwardRef(function Alert(
  { children, ...props }: AccordionProps,
  ref
) {
  return (
    <Accordion {...props} ref={ref}>
      {children}
    </Accordion>
  );
});

export default MainAccordion;
