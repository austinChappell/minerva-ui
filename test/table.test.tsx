import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import {
  Table,
  TableHeader,
  TableRow,
  TableCell,
  TableBody,
  TableHeaderCell,
} from '../src';
import { ThemeProvider } from '../src';

const BasicTable = () => {
  const ref = React.useRef(null);

  return (
    <ThemeProvider>
      <Table ref={ref}>
        <TableHeader>
          <TableRow>
            <TableHeaderCell>Name</TableHeaderCell>
            <TableHeaderCell>Job</TableHeaderCell>
            <TableHeaderCell>Status</TableHeaderCell>
            <TableHeaderCell>Role</TableHeaderCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          <TableRow>
            <TableCell>Test</TableCell>
            <TableCell>Tester</TableCell>
            <TableCell>Testing</TableCell>
            <TableCell>Test Master</TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </ThemeProvider>
  );
};

describe('<Table />', () => {
  it('should render', () => {
    const { container } = render(<BasicTable />);
    expect(container).toMatchSnapshot();
  });

  it('should render', () => {
    const { container } = render(<BasicTable />);

    expect(container).toMatchSnapshot();
  });

  // it('should change style if disabled', () => {
  //   const { container, getByRole } = render(
  //     <ThemeProvider>
  //       <Table disabled>Disabled Table</Table>
  //     </ThemeProvider>
  //   );

  //   expect(container).toMatchSnapshot();
  //   expect(getByRole('button')).toHaveStyleRule('opacity', '0.4', {
  //     modifier: ':disabled',
  //   });
  // });
});
