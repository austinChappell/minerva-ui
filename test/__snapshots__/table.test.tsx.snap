// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render 1`] = `
.c0 {
  border-radius: 8px;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.1),0 1px 2px 0 rgba(0,0,0,0.06);
}

.c1 {
  min-width: 100%;
  border-collapse: collapse;
}

.c2 {
  padding: 12px 24px;
  color: #6b7280;
  background-color: #f9fafb;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 12px;
  -webkit-letter-spacing: 0.6px;
  -moz-letter-spacing: 0.6px;
  -ms-letter-spacing: 0.6px;
  letter-spacing: 0.6px;
  text-align: left;
  border-bottom: 1px solid #e5e7eb;
}

.c3 {
  padding: 16px 24px;
  color: #161e2e;
  background-color: #fff;
  font-size: 14px;
  border-bottom: 1px solid #e5e7eb;
}

<div>
  <div
    class="c0"
  >
    <table
      class="c1"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class="c2"
          >
            Name
          </th>
          <th
            class="c2"
          >
            Job
          </th>
          <th
            class="c2"
          >
            Status
          </th>
          <th
            class="c2"
          >
            Role
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
        >
          <td
            class="c3"
          >
            Test
          </td>
          <td
            class="c3"
          >
            Tester
          </td>
          <td
            class="c3"
          >
            Testing
          </td>
          <td
            class="c3"
          >
            Test Master
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<Table /> should render 2`] = `
.c0 {
  border-radius: 8px;
  box-shadow: 0 1px 3px 0 rgba(0,0,0,0.1),0 1px 2px 0 rgba(0,0,0,0.06);
}

.c1 {
  min-width: 100%;
  border-collapse: collapse;
}

.c2 {
  padding: 12px 24px;
  color: #6b7280;
  background-color: #f9fafb;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 12px;
  -webkit-letter-spacing: 0.6px;
  -moz-letter-spacing: 0.6px;
  -ms-letter-spacing: 0.6px;
  letter-spacing: 0.6px;
  text-align: left;
  border-bottom: 1px solid #e5e7eb;
}

.c3 {
  padding: 16px 24px;
  color: #161e2e;
  background-color: #fff;
  font-size: 14px;
  border-bottom: 1px solid #e5e7eb;
}

<div>
  <div
    class="c0"
  >
    <table
      class="c1"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class="c2"
          >
            Name
          </th>
          <th
            class="c2"
          >
            Job
          </th>
          <th
            class="c2"
          >
            Status
          </th>
          <th
            class="c2"
          >
            Role
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
        >
          <td
            class="c3"
          >
            Test
          </td>
          <td
            class="c3"
          >
            Tester
          </td>
          <td
            class="c3"
          >
            Testing
          </td>
          <td
            class="c3"
          >
            Test Master
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
